File: TripKitAndroid/src/main/java/com/skedgo/android/tripkit/RegionFinder.java
package com.skedgo.android.tripkit;

import com.skedgo.android.common.model.Region;
import com.skedgo.android.common.util.PolyUtil;
import com.skedgo.android.common.util.TripKitLatLng;

import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

class RegionFinder {
  private final Map<String, List<TripKitLatLng>> polygonCache = new ConcurrentHashMap<>();

  public boolean contains(Region region, double lat, double lng) {
    final List<TripKitLatLng> polygon = getPolygon(region, polygonCache);
    return polygon != null
        && PolyUtil.containsLocation(lat, lng, polygon, true);
  }

  public void invalidate() {
    polygonCache.clear();
  }

  private List<TripKitLatLng> getPolygon(Region region, Map<String, List<TripKitLatLng>> polygonCache) {
    final String name = region.getName();
    List<TripKitLatLng> polygon = polygonCache.get(name);
    if (polygon == null) {
      polygon = PolyUtil.decode(region.getEncodedPolyline());
      polygonCache.put(name, polygon);
    }
    return polygon;
  }
}

File: TripKitAndroid/src/main/java/com/skedgo/tripkit/RegionFinder.kt
package com.skedgo.tripkit

import com.skedgo.tripkit.common.model.region.Region
import com.skedgo.tripkit.common.util.PolyUtil
import com.skedgo.tripkit.common.util.TripKitLatLng
import java.util.concurrent.ConcurrentHashMap

class RegionFinder {

    private val polygonCache: MutableMap<String, List<TripKitLatLng>> = ConcurrentHashMap()

    fun contains(region: Region, lat: Double, lng: Double): Boolean {
        val polygon = getPolygon(region, polygonCache)
        return polygon != null && PolyUtil.containsLocation(lat, lng, polygon, true)
    }

    fun invalidate() {
        polygonCache.clear()
    }

    private fun getPolygon(
        region: Region,
        polygonCache: MutableMap<String, List<TripKitLatLng>>
    ): List<TripKitLatLng>? {
        val name = region.name.orEmpty()
        var polygon = polygonCache[name]
        if (polygon == null) {
            polygon = PolyUtil.decode(region.encodedPolyline)
            polygonCache[name] = polygon
        }
        return polygon
    }
}
